collector/hwmon_linux.go

```diff
@@ -0,0 +1,392 @@
+// Copyright 2016 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// +build !nohwmon
+
+package collector
+
+import (
+	"errors"
+	"io/ioutil"
+	"os"
+	"path"
+	"path/filepath"
+	"regexp"
+	"strconv"
+	"strings"
+
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+const (
+	hwMonSubsystem = "hwmon"
+)
+
+type hwMonCollector struct{}
+
+var (
+	invalidMetricChars *regexp.Regexp
+	labelDesc          = []string{"hwmon", "sensor", "label"}
+)
+
+func init() {
+	Factories["hwmon"] = NewhwMonCollector
+	invalidMetricChars, _ = regexp.Compile("[^a-z0-9:_]")
+}
+
+// Takes a prometheus registry and returns a new Collector exposing
+// /sys/class/hwmon stats (similar to lm-sensors).
+func NewhwMonCollector() (Collector, error) {
+	return &hwMonCollector{}, nil
+}
+
+func cleanMetricName(name string) string {
+	lower := strings.ToLower(name)
+	replaced := invalidMetricChars.ReplaceAllLiteralString(lower, "_")
+	cleaned := strings.Trim(replaced, "_")
+	return cleaned
+}
+
+func findLabel(name string, labels map[string]string) string {
+	if result, ok := labels[name]; ok {
+		return result
+	}
+	splitPosition := strings.LastIndex(name, "_")
+	if splitPosition == -1 {
+		return ""
+	}
+	result := findLabel(name[0:splitPosition], labels)
+	if result == "" {
+		return ""
+	}
+	return result + name[splitPosition:]
+}
+
+func addValueFile(data map[string]map[string]string, sensor string, prop string, file string) {
+	raw, e := ioutil.ReadFile(file)
+	if e != nil {
+		return
+	}
+	value := strings.Trim(string(raw), "\n")
+
+	if _, ok := data[sensor]; !ok {
+		data[sensor] = make(map[string]string)
+	}
+
+	data[sensor][prop] = value
+}
+
+// maps a sensor file name, e.g. cpu0_vid, to a sensor type, e.g. cpu
+func extractSensorType(filename string) string {
+	nameFragments := strings.Split(filename, "_")
+	sensor := nameFragments[0]
+	numlen := 0
+	for _, runeValue := range sensor {
+		if runeValue < '0' || runeValue > '9' {
+			numlen = 0
+		} else {
+			numlen++
+		}
+	}
+	if numlen == 0 {
+		return sensor
+	}
+	return sensor[0 : len(sensor)-numlen]
+}
+
+func collectSensorData(dir string, data map[string]map[string]string) (err error) {
+	sensorFiles, dirError := ioutil.ReadDir(dir)
+	if dirError != nil {
+		return dirError
+	}
+	for _, file := range sensorFiles {
+		filename := file.Name()
+		if filename == "vrm" || filename == "beep_enable" || filename == "update_interval" {
+			// the only filenames without a number
```

Comment:
well, those are all sensor files that don't follow the schema
```
<sensortype>[0-9]*_<field>
```

This comes from https://www.kernel.org/doc/Documentation/hwmon/sysfs-interface and thus won't change quickly. I could move this to a list, though